include/ibamr/LevelSetUtilities.h

```diff
@@ -0,0 +1,457 @@
+              d_tag_max_value(tag_max_value)
+        {
+            // intentionally left blank
+        } // TagLevelSetRefinementCells
+
+        LevelSetContainer getLevelSetContainer() const
+        {
+            return d_ls_container;
+        } // getLevelSetContainer
+
+        void setTagMinValue(double tag_min)
+        {
+            d_tag_min_value = tag_min;
+            return;
+        } // setTagMinValue
+
+        double getTagMinValue() const
+        {
+            return d_tag_min_value;
+        } // getTagMinValue
+
+        void setTagMaxValue(double tag_max)
+        {
+            d_tag_max_value = tag_max;
+            return;
+        } // setTagMaxValue
+
+        double getTagMaxValue() const
+        {
+            return d_tag_max_value;
+        } // getTagMaxValue
+
+    private:
+        LevelSetContainer d_ls_container;
+        double d_tag_min_value = 0.0;
+        double d_tag_max_value = 0.0;
+    }; // TagLevelSetRefinementCells
+
+    /*!
+     * \brief Preprocessing call back function to be hooked into IBAMR::HierarchyIntegrator class
+     * to tag the cells for grid refinement based on the given tagging criteria.
+     *
+     * This static member should be registered with an appropriate hierarchy integrator
+     * via registerApplyGradientDetectorCallback().
+     *
+     * \param ctx is the pointer to the TagLSRefinementCells class object.
+     */
+    void tagLSCells(SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM> > hierarchy,
+                    const int level_number,
+                    const double error_data_time,
+                    const int tag_index,
+                    const bool initial_time,
+                    const bool uses_richardson_extrapolation_too,
+                    void* ctx);
+
+    /*!
+     * \brief A lightweight class that stores the current value of the Lagrange multiplier
+     *  for the level set variable.
+     */
+    class LevelSetMassLossFixer : public SAMRAI::tbox::Serializable
+    {
+    public:
+        /*!
+         * \brief Constructor of the class.
+         * @param input_db provides parameters such as enable_logging, correction_interval, max_its, rel_tol, and
+         * half_width.
+         */
+        LevelSetMassLossFixer(std::string object_name,
+                              SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_integrator,
+                              std::vector<SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > > ls_vars,
+                              SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db =
+                                  SAMRAI::tbox::Pointer<SAMRAI::tbox::Database>(nullptr),
+                              bool register_for_restart = true);
+
+        ~LevelSetMassLossFixer();
+
+        void putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db) override;
+
+        LevelSetContainer getLevelSetContainer() const
```

Comment:
Return a `const` reference, and let the user decide if it needs to be copied.
```suggestion
        const LevelSetContainer& getLevelSetContainer() const
```

Reply:
Hi Aaron, If we return a ```const``` reference, then it is not possible to call non-const member functions like ```setInterfaceHalfWidth(double ncells)``` in ```LevelSetContainer``` class using this reference and set the ```ncells``` value.

Reply:
Please ignore my previous comment. Sorry, I did not understand it correctly.

Reply:
I'm not sure I understand. If you return a copy, you also can't modify the underlying `LevelSetContainer`.

But more generally, by making `LevelSetMassLossFixer` contain a `LevelSetContainer`, you are saying that the `LevelSetMassLossFixer` is acting on a particular `LevelSetContainer`. You shouldn't be treating the `LevelSetMassLossFixer` as a `LevelSetContainer`, and `LevelSetMassLossFixer` shouldn't have to modify the underlying container.

Reply:
Sorry for the bug. Now I return the const reference from getLevelSetContainer(). To set the ```d_ncells```, I have created a new function ```void setLevelSetContainerInterfaceHalfWidth(double ncells)``` inside the ```LevelSetMassLossFixer```. Please take a look. I have run clang-format already. But for some reason, I think it is messed up. That is why the CI is failing. I am not sure how to fix this.